lib/remap-functions/src/parse_key_value.rs

```diff
@@ -0,0 +1,370 @@
+                    (
+                        "@timestamp".to_string(),
+                        "2020-12-31T12:43:22.2322232Z".into()
+                    ),
+                    ("key#hash".to_string(), "value".into()),
+                    ("key=with=special=characters".to_string(), "value".into()),
+                    ("key".to_string(), "with special=characters".into()),
+                ]
+            )),
+            parse_line(
+                r#"ook=pook @timestamp=2020-12-31T12:43:22.2322232Z key#hash=value "key=with=special=characters"=value key="with special=characters""#,
+                "=",
+                " "
+            )
+        );
+    }
+
+    #[test]
+    fn test_parse_value() {
+        // delimited
+        assert_eq!(Ok(("", "noog".into())), parse_value(" ")(r#""noog""#));
+
+        // undelimited
+        assert_eq!(Ok(("", "noog".into())), parse_value(" ")("noog"));
+    }
+
+    test_type_def![
+        value_string {
+            expr: |_| ParseKeyValueFn {
+                value: Literal::from("foo").boxed(),
+                key_value_delimiter: None,
+                field_delimiter: None,
+            },
+            def: TypeDef {
+                fallible: true,
+                kind: Kind::Map,
+                ..Default::default()
+            },
+        }
+
+        value_non_string {
+            expr: |_| ParseKeyValueFn {
+                value: Literal::from(1).boxed(),
+                key_value_delimiter: None,
+                field_delimiter: None,
+            },
+            def: TypeDef {
+                fallible: true,
+                kind: Kind::Map,
+                ..Default::default()
+            },
+        }
+
+        optional_value_string {
+            expr: |_| ParseKeyValueFn {
+                value: Literal::from("ook").boxed(),
+                key_value_delimiter: Some(Literal::from("=").boxed()),
+                field_delimiter: None,
+            },
+            def: TypeDef {
+                fallible: true,
+                kind: Kind::Map,
+                ..Default::default()
+            },
+        }
+
+        optional_value_non_string {
+            expr: |_| ParseKeyValueFn {
+                value: Literal::from("ook").boxed(),
+                key_value_delimiter: Some(Literal::from(1).boxed()),
+                field_delimiter: None,
+            },
+            def: TypeDef {
+                fallible: true,
+                kind: Kind::Map,
+                ..Default::default()
+            },
+        }
+    ];
+
+    test_function![
+        parse_key_value => ParseKeyValue;
+
+        default {
+            args: func_args! [
+                value: r#"at=info method=GET path=/ host=myapp.herokuapp.com request_id=8601b555-6a83-4c12-8269-97c8e32cdb22 fwd="204.204.204.204" dyno=web.1 connect=1ms service=18ms status=200 bytes=13 tls_version=tls1.1 protocol=http"#,
+            ],
+            want: Ok(value!({"at": "info",
```

Comment:
nit/remark: again, just mentioning for future reference: the map notation also allows keys without quotes (as long as they are a single identifier).

This can sometimes make maps with lots of string values easier to parse/read for others.

E.g.:

`{ "foo": "bar" }` vs `{ foo: "bar" }` (the difference is more profound with syntax highlighting)